### ISSUE TITLE
The command "goi18n constants" is missing

### ISSUE THREAD

**Issue by VadimKulagin** [nicksnyder/go-i18n#137]
Hello! 
I use "github.com/nicksnyder/go-i18n/v2/i18n" and I want to use the command "goi18n constants", but for some reason it is unavailable in the 2nd version of this package.

I tried to use the first version of the package to generate constants, but it seems that it is incompatible with the format of the 2nd version:

```
goi18n constants es.toml
failed to load translation file es.toml because unable to parse translation #0 because invalid plural category description
map[id:PersonCats translation:map[description:The number of cats a person has one:{{.Name}} has {{.Count}} cat. other:{{.Name}} has {{.Count}} cats.]]
```

```
# es.toml
[PersonCats]
description = "The number of cats a person has"
one = "{{.Name}} has {{.Count}} cat."
other = "{{.Name}} has {{.Count}} cats."
```

Can you explain why this command is not available in the second version of the package and do you have plans to return it?

**Comment by nicksnyder** [comment]
Most of the API changed in v2 and I did not implement the constants command because it wasn't clear if it would be necessary/useful any more.

In v2, the recommended approach is to store source translations in your go code and then extract the translations to message files. Since your Go code is the source of truth, it doesn't make much sense to me to want to generate constants.

**Comment by VadimKulagin** [comment]
Constants have several advantages:
* with their help, you can easily find which messages are not used in the application at all (GoLand highlights such constants as unused);
* with the help of constants it is easy to find all the uses of a particular message, especially if it has a frequently used key, such as "id", "name", "value", etc;
* they allow you to immediately detect the change / disappearance of the message key (in the message file), since the application code will refer to a constant that is no longer there.

And about the `goi18n extract`. This command does not find any messages in the project at all and creates an empty file `active.en.toml`. Here is an example of code that this command for some reason does not see:
```go
package main

import (
	"github.com/nicksnyder/go-i18n/v2/i18n"
	"golang.org/x/text/language"
	"gopkg.in/yaml.v2"
	"log"
)

func main() {
	bundle := &i18n.Bundle{
		DefaultLanguage: language.English,
	}
	bundle.RegisterUnmarshalFunc("yaml", yaml.Unmarshal)
	bundle.MustLoadMessageFile("messages/active.en.yaml")
	bundle.MustLoadMessageFile("messages/active.ru.yaml")

	loc := i18n.NewLocalizer(bundle, "en")

	str, err := loc.Localize(&i18n.LocalizeConfig{
		MessageID:   "LANG",
	})

	log.Printf("str: %v, err %+v",  str, err)
}
```

And even if this command could parse such code, I doubt that it would be able to find the `Localize` calls if they are executed through some kind of wrapper. For example, I made a separate wrapper that has additional methods for working with `Localize`.

**Comment by nicksnyder** [comment]
I understand the value of constants in v1. Like v1, v2 allows you to make localize calls using only IDs too and I can see that if you use it in this way a constants command would probably be useful for all the same reasons as it was useful in v1.

The extract command should be able to extract that "LANG" message ID. I haven't had a chance to see what the specific problem is here. You can look at the [code](http://localhost:3080/github.com/nicksnyder/go-i18n/-/blob/v2/goi18n/extract_command.go#L141) and [tests](http://localhost:3080/github.com/nicksnyder/go-i18n/-/blob/v2/goi18n/extract_command_test.go).

**Comment by VadimKulagin** [comment]
> The extract command should be able to extract that "LANG" message ID.

Unfortunately, it does not.

Over the weekend, I'll look at the goi18n code to find the problem.

**Comment by VadimKulagin** [comment]
Thanks for the detailed answer!

> I am curious to know why you don't want to put source translations in Go files

From my experience with other projects, I am used to the fact that localization texts should be stored separately from code. And after reading the documentation for the first version of this package, I decided that using constants is close in spirit to my previous experience.

However, thank you very much for pointing out that the message extractor is looking for `&i18n.Message` in the code and that I can store the messages in a separate package almost the same as constants:

```go
package R

import "github.com/nicksnyder/go-i18n/v2/i18n"

// Messages
var (
	Example = &i18n.Message{
		ID:    "Example",
		Other: "Some Example",
	}
	Seconds = &i18n.Message{
		ID:    "Seconds",
		One:   "{{.N}} second",
		Other: "{{.N}} seconds",
	}
)
```

And in the end, I can easily access them through my Localizer wrapper:
```go
localizer.Message(R.Example)
// or
localizer.Plural(R.Seconds, secondsValue)
// or
localizer.MessageData(R.Seconds, i18n.Data{"N": secondsValue})
```

This turned out to be a quite convenient workflow. And this made it possible to get rid of two additional files with translations `en.yaml` & `ru.yaml`, leaving only `active.en.yaml` & `active.ru.yaml`.

I think you should add such usage examples to the documentation. It is also worth adding convenient methods for localizing messages in the plural and with passing parameters (as I showed in my example).

And let me ask one more question: what is the need for the `hash` field in the translation file and do I need to transfer it from `translate.<LANG>.yaml` to `active.<LANG>.yaml`?

**Comment by nicksnyder** [comment]
> This turned out to be a quite convenient workflow.

I am glad that this worked well for you. It sounds like you no longer have a need for the constants command for v2?

> I think you should add such usage examples to the documentation.

Can you be a little bit more specific about what kind of documentation would have helped? Maybe I can add a migration guide for v1 to v2?

> It is also worth adding convenient methods for localizing messages in the plural and with passing parameters (as I showed in my example).

Yeah, I can consider adding some convenience methods.

> what is the need for the `hash` field in the translation file and do I need to transfer it from `translate.<LANG>.yaml` to `active.<LANG>.yaml`?

The hash field in the translation file uniquely identifies the content of the source translation. This is used to detect if the source translation has changed between when it was extracted and when the translation came back. The hash field is not necessary in `active` files.

https://github.com/nicksnyder/go-i18n/blob/master/v2/internal/message.go#L13-L15

**Comment by nicksnyder** [comment]
Also, given the above info and your new workflow, do you still need https://github.com/nicksnyder/go-i18n/pull/141 ?

**Comment by VadimKulagin** [comment]
> It sounds like you no longer have a need for the constants command for v2?

Yes. Now I store the Message variables in a separate package and use them instead of constants.

> Can you be a little bit more specific about what kind of documentation would have helped?

I mean, it's worth adding examples to illustrate that a user can store messages in a separate package (for example, `msg` or `R`):

```go
package msg

import "github.com/nicksnyder/go-i18n/v2/i18n"

var (
	Example = &i18n.Message{
		ID:    "Example",
		Other: "Some Example",
	}
	Seconds = &i18n.Message{
		ID:    "Seconds",
		One:   "{{.N}} second",
		Other: "{{.N}} seconds",
	}
)
```

And that these messages can then be used in one simple line:
```go
localizer.MustLocalize(&i18n.LocalizeConfig{DefaultMessage: msg.Example})
```

Or even easier if you add a method to localize the passed *i18n.Message:
```go
localizer.Message(msg.Example)
```

It is much more convenient to use instead of such bulky structures:
```go
	localizer.MustLocalize(&i18n.LocalizeConfig{
		DefaultMessage: &i18n.Message{
			ID:    "Example",
			Other: "Some Example",
		},
	}))
```

And then it is worth mentioning that to translate all these messages, the user needs to run `goi18n extract` and perform the [necessary actions](https://github.com/nicksnyder/go-i18n#translating-new-messages).

And it is worth mentioning in the documentation what you wrote above:

> The extract command looks for i18n.Message and i18n.LocalizeConfig struct literals, so it doesn't matter what helper functions you create, as long as you pass through those literals.

This is a good hint for those who plan to use `goi18n extract` and want to understand how it works and what to expect from this utility without digging the code of this project.

> Maybe I can add a migration guide for v1 to v2?

I have not used the first version of the package. I just used the goi18n (v1) utility to extract constants and use them with the v2 package. But I think that a migration guide from v1 to v2 will be useful to someone.

> Also, given the above info and your new workflow, do you still need #141 ?

It seems that initially I did not fully understand how `goi18n extract` works. I expected that having a file with messages `en.yaml` and having such an expression in the code:

```go
	loc.Localize(&i18n.LocalizeConfig{
		MessageID:   "LANG",
	})
```

I thought that after running `goi18n extract`, I would get a file `active.en.yaml` containing only those messages from `en.yaml` whose MessageID is mentioned in the code 😄 

It seems that such confusion came out because I didn’t use this package before and messed up something between the v1 and v2 documentation.

So I doubt that I need changes from #141.

**Comment by nicksnyder** [comment]
> it's worth adding examples to illustrate that a user can store messages in a separate package 

I feel like this shouldn't be necessary since it is a general programming practice (not knowledge specific to this library) to use variables to reduce verbosity of a call site if desired. This is certainly necessary if the same message is shared in multiple locations.

> Or even easier if you add a method to localize the passed *i18n.Message:

Added in https://github.com/nicksnyder/go-i18n/pull/173

> And then it is worth mentioning that to translate all these messages, the user needs to run `goi18n extract` and perform the necessary actions.

This is documented in the main readme (which you linked to), so it is unclear to me what you are suggesting to improve here.

> This is a good hint for those who plan to use goi18n extract and want to understand how it works and what to expect from this utility without digging the code of this project.

I added improved readme in https://github.com/nicksnyder/go-i18n/commit/3b004414b00363adb329d3dd6ae8cb561aa43b08#diff-04c6e90faac2675aa89e2176d2eec7d8R70

> It seems that such confusion came out because I didn’t use this package before and messed up something between the v1 and v2 documentation.

I will try to remove v1 from master while maintaining backward compatibility to avoid confusions like this in the future:
https://github.com/nicksnyder/go-i18n/issues/172

I am going to close this issue since we seem to have resolved the initial discussion about the constants command. Feel free to open separate issues as you find them or open PRs to suggest improvements to APIs and/or doc.
